lucet-runtime/lucet-runtime-internals/src/module/dl.rs

```diff
@@ -117,6 +118,94 @@ impl DlModule {
+            use nix::sys::mman::ProtFlags;
+
+            // Yoinked from region/mmap.rs for now
+            unsafe fn mprotect(
+                addr: *mut c_void,
+                length: libc::size_t,
+                prot: ProtFlags,
+            ) -> nix::Result<()> {
+                nix::errno::Errno::result(libc::mprotect(addr, length, prot.bits())).map(drop)
+            }
+
+            let runtime = unix::Library::this();
+
+            let manifest_page_offset =
+                function_manifest.as_ptr() as u64 % alloc::host_page_size() as u64;
+            let manifest_page_ptr =
+                (function_manifest.as_ptr() as u64 - manifest_page_offset) as *mut c_void;
+
+            // Because the function manifest may not be page-aligned, the manifest may actually
+            // start far past 0, with pages laid out like something like:
+            //
+            // 0                X                   Y
+            // | data data data | function manifest |
+            //
+            // where Y-X == manifest length as bytes, but X can be anywhere in the PAGE_SIZE bytes
+            // after 0. Since mprotect will start at 0 from whatever address, we must actually
+            // adjust protections for the whole span [0..Y). We do this by specifying that region
+            // as the length to mprotect.
+            let manifest_region_len = (manifest_page_offset
+                + (function_manifest.len() * std::mem::size_of::<FunctionSpec>()) as u64)
+                as usize;
+
+            mprotect(
+                manifest_page_ptr,
+                manifest_region_len,
+                ProtFlags::PROT_READ | ProtFlags::PROT_WRITE,
+            )?;
+
+            for import in module_data.import_functions() {
+                let host_sym = match import.mapped_to {
+                    Some(sym) => sym,
+                    None => {
+                        return Err(lucet_incorrect_module!(
+                            "unbound symbol cannot be resolved: {}/{}",
+                            import.module,
+                            import.name
+                        ));
+                    }
+                };
+
+                let ptr = runtime
+                    .get::<extern "C" fn()>(host_sym.as_bytes())
+                    .map_err(Error::DlError)?;
+
+                let existing_entry = function_manifest
+                    .get_mut(import.fn_idx.as_u32() as usize)
+                    .unwrap();
+
+                // TODO: if this import is ever resolved from another WASM module, we can
+                // read trap and code length information from that module. In fact, this might even
+                // be rewritable as a replace by a clone of the exported FunctionSpec that satisfies
+                // this import!
+                let new_entry = FunctionSpec::new(
+                    *ptr as u64,
+                    existing_entry.code_len(),
+                    existing_entry
```

Comment:
what is the situation where it'd be possible for an import function to have associated trap info?

Reply:
admittedly I'm getting ahead of myself here, these should all be 0 when we're updating them at load-time. Eventually if we import from another module, we'd want the trap information available here to look up in the case something goes sideways.